core/modules/ccontrol/testCControl.cc

```diff
@@ -619,36 +622,82 @@ BOOST_DATA_TEST_CASE(antlr_compare, QUERIES, query) {
-    // Instead, we test NOT LIKE purely with the antlr4 parser, by generating a statement with LIKE, and
-    // we then manually change that IR to be NOT LIKE, and compare that with a generated NOT LIKE statement.
-    // We know the LIKE statement is correct because we have a test that compares it with the
-    // antlr2-generated LIKE statement, in antlr_compare.
-    auto parserWithNot = parser::SelectParser::newInstance(
-            "select shortName from Filter where shortName NOT LIKE 'Z'", parser::SelectParser::ANTLR4);
-    BOOST_REQUIRE_NO_THROW(parserWithNot->setup());
-    auto selectStatementWithNot = parserWithNot->getSelectStmt();
-    BOOST_REQUIRE(selectStatementWithNot != nullptr);
-
-    auto parserWithoutNot = parser::SelectParser::newInstance(
-            "select shortName from Filter where shortName LIKE 'Z'", parser::SelectParser::ANTLR4);
-    BOOST_REQUIRE_NO_THROW(parserWithoutNot->setup());
-    auto selectStatementWithoutNot = parserWithoutNot->getSelectStmt();
-    BOOST_REQUIRE(selectStatementWithoutNot != nullptr);
-
-    // flip the 'not' on the 'without' statement to make it 'with not'
-    auto whereClauseRef = selectStatementWithoutNot->getWhereClause();
-    auto orTerm = std::dynamic_pointer_cast<query::OrTerm>(whereClauseRef.getRootTerm());
-    auto andTerm = std::dynamic_pointer_cast<query::AndTerm>(orTerm->_terms[0]);
-    auto boolFactor = std::dynamic_pointer_cast<query::BoolFactor>(andTerm->_terms[0]);
-    auto likePredicate = std::dynamic_pointer_cast<query::LikePredicate>(boolFactor->_terms[0]);
-    likePredicate->hasNot = true;
-    // now the the LIKE and the NOT LIKE selectStatements should be exactly the same:
-    BOOST_REQUIRE_EQUAL(*selectStatementWithNot, *selectStatementWithoutNot);
+// in many cases we can use the antlr2-parser-generated IR to test the IR generated by the antlr4 parser.
+// This requires that the query can be parsed by our antlr2 parser code, and as we add query coverage
+// to our antlr4 parser code we will not be able to use the antlr2 IR for testing.
+// Instead, we test these queries purely with the antlr4 parser, by generating a similar statement and we
+// then manually change that IR to be like the expected IR, and compare that with the IR generated by the
+// test statement.
+// We know the changed test statement IR is correct because we have a test that compares it with the
+// antlr2-generated statement, in antlr_compare.
+
+struct Antlr4CompareQueries {
+    Antlr4CompareQueries(std::string const & iQuery, std::string const & iCompQuery,
+            std::function<void(query::SelectStmt::Ptr const&)> const & iModFunc,
+            std::string const & iSerializedQuery)
+        : query(iQuery)
+        , compQuery(iCompQuery)
+        , serializedQuery(iSerializedQuery)
+        , modFunc(iModFunc)
+    {}
+
+    std::string query;
+    std::string compQuery;
+    std::string serializedQuery;
+    std::function<void(query::SelectStmt::Ptr const)> modFunc;
```

Comment:
Some comments would probably make this easier to understand. 'query' - query to test. 'compQuery' - antr2 comparison query. 'serializedQuery' - expected reconstruction.  'modFunc' -  makes the changes to compQuery IR so it should be equivalent to 'query'.

Reply:
close, but not quite (compQuery will be parsed by antlr4, not antlr2, and then modified by modFunc, as you observed). So, point proven :-) I'll add comments.